database/agents.go

```diff
@@ -117,8 +117,14 @@ func (db *DB) InsertAgent(agt mig.Agent) (err error) {
 // UpdateAgentHeartbeat updates the heartbeat timestamp of an agent in the database
 // unless the agent has been marked as destroyed or upgraded
 func (db *DB) UpdateAgentHeartbeat(agt mig.Agent) (err error) {
-	_, err = db.c.Exec(`UPDATE agents SET status=$1, heartbeattime=$2 WHERE id=$3`,
-		mig.AgtStatusOnline, agt.HeartBeatTS, agt.ID)
+	jEnv, err := json.Marshal(agt.Env)
+	if err != nil {
+		err = fmt.Errorf("Failed to marshal agent environment: '%v'", err)
+		return
+	}
+	_, err = db.c.Exec(`UPDATE agents SET status=$1, heartbeattime=$2,
+		environment=$3, name=$4 WHERE id=$5`,
+		mig.AgtStatusOnline, agt.HeartBeatTS, jEnv, agt.Name, agt.ID)
```

Comment:
@jvehent Yeah. We could mark the agent immediately as offline on update, or potentially introduce a new status (e.g., "superseded") for this. Since we search by queue and PID and != offline don't think we can just introduce a new row here alone. Going immediately to offline I don't really like as it bypasses the normal agent lifecycle, but at that point maybe its the correct status for the agent.

Another option is adding a new relation to the agents table, to track env/tag/name over time and removing those columns from agents, this seems more complicated than is required though.